File: backend/api/src/connectors/facebook.ts
import fetch from 'node-fetch';
import * as DataLoader from 'dataloader';
import * as winston from 'winston';

import { FacebookConfig } from '../config'

const FACEBOOK_GRAPH_ROOT = 'https://graph.facebook.com/v2.8';

export interface FacebookQuery {
    path: string;
    options?: {[key: string]: string }
}

export class FacebookConnector {
    loader: DataLoader<FacebookQuery, any>;

    constructor(public config: FacebookConfig) {
        this.loader = new DataLoader<FacebookQuery,any>((queries) => {
            return Promise.all(queries.map(query => {
                let paramString=`access_token=${ config.appID }|${ config.appSecret }`
                for(const option in query.options){
                    const queryString = `${ option }=${ query.options[option] }`;
                        paramString = `${ paramString }&${ queryString }`
                }
                const url = `${ FACEBOOK_GRAPH_ROOT }/${ query.path }?${ paramString }`;
                return fetch(url).then(data => data.json());
            }));
        }, { batch: false })
    }

    get(query: FacebookQuery) {
        return this.loader.load(query);
    }
}

File: backend/src/util/facebook.py
import logging
import requests

from django.conf import settings

FACEBOOK_GRAPH_ROOT = 'https://graph.facebook.com/v2.8'

logger = logging.getLogger('showdown.%s' % __name__)

def get_facebook_results(path, options=None):
    facebook_config = settings.FACEBOOK
    #access_token = '%(app_id)s|%(app_secret)s' % facebook_config

    if not options: options = {}

    #options['access_token'] = access_token
    url = '%s/%s' % (FACEBOOK_GRAPH_ROOT, path)
    response = requests.get(url, params=options)

    return response.json()

def get_facebook_profile(user_id):
    return get_facebook_results(path=user_id)

def get_token_info(accessToken):
    result = get_facebook_results(
        path='me', 
        options={'access_token': accessToken}
    )
    return result['id'], result['name']
